dali/kernels/imgproc/sampler.h

```diff
@@ -199,83 +217,161 @@ struct Sampler<DALI_INTERP_LINEAR, In> {
+      In s00 = NN.at(ivec2(x0,   y0),   c, border_value);
+      In s01 = NN.at(ivec2(x0+1, y0),   c, border_value);
+      In s10 = NN.at(ivec2(x0,   y0+1), c, border_value);
+      In s11 = NN.at(ivec2(x0+1, y0+1), c, border_value);
       float s0 = s00 * px + s01 * qx;
       float s1 = s10 * px + s11 * qx;
       out_pixel[c] = ConvertSat<T>(s0 + (s1 - s0) * qy);
     }
   }
 
   template <typename T, typename BorderValue>
-  DALI_HOST_DEV void operator()(
-      T *pixel,
-      int x, int y, int c,
-      BorderValue border_value) const {
-    Sampler<DALI_INTERP_NN, In> s(surface);
-    s(pixel, x, y, c, border_value);
+  DALI_HOST_DEV DALI_FORCEINLINE void operator()(
+    T *out_pixel, vec2 pos, int c, BorderValue border_value) const {
+    out_pixel[c] = at<T>(pos, c, border_value);
   }
 
   template <typename T, typename BorderValue>
-  DALI_HOST_DEV void operator()(
-      T *pixel,
-      int x, int y,
-      BorderValue border_value) const {
-    Sampler<DALI_INTERP_NN, In> s(surface);
-    s(pixel, x, y, border_value);
+  DALI_HOST_DEV DALI_FORCEINLINE void operator()(
+    T *out_pixel, ivec2 pos, int c, BorderValue border_value) const {
+    out_pixel[c] = at<T>(pos, c, border_value);
   }
 
   template <typename T, typename BorderValue>
-  DALI_HOST_DEV void
-  operator()(T *pixel, ivec2 xy, int c, BorderValue border_value) const {
-    (*this)(pixel, xy.x, xy.y, c, border_value);
+  DALI_HOST_DEV DALI_FORCEINLINE void operator()(
+    T *out_pixel, ivec2 pos, BorderValue border_value) const {
+    Sampler2D<DALI_INTERP_NN, In> NN(surface);
+    NN(out_pixel, pos, border_value);
+  }
+};
+
+template <typename In>
+struct Sampler<DALI_INTERP_LINEAR, 3, In> {
+  static constexpr int spatial_ndim = 3;
+  using icoords = ivec3;
+  using fcoords = vec3;
+
+  Sampler() = default;
+  DALI_HOST_DEV explicit Sampler(const Surface3D<const In> &surface)
+  : surface(surface) {}
+
+  Surface3D<const In> surface;
+
+  template <typename T = In, typename BorderValue>
+  DALI_HOST_DEV T at(
+      vec3 pos, int c, BorderValue border_value) const {
+    Sampler3D<DALI_INTERP_NN, In> NN(surface);
+    float x = pos.x - 0.5f;
+    float y = pos.y - 0.5f;
+    float z = pos.z - 0.5f;
+    int x0 = floor_int(x);
+    int y0 = floor_int(y);
+    int z0 = floor_int(z);
+    In s000 = NN.at(ivec3(x0,   y0,   z0),   c, border_value);
+    In s001 = NN.at(ivec3(x0+1, y0,   z0),   c, border_value);
+    In s010 = NN.at(ivec3(x0,   y0+1, z0),   c, border_value);
+    In s011 = NN.at(ivec3(x0+1, y0+1, z0),   c, border_value);
+    In s100 = NN.at(ivec3(x0,   y0,   z0+1), c, border_value);
+    In s101 = NN.at(ivec3(x0+1, y0,   z0+1), c, border_value);
+    In s110 = NN.at(ivec3(x0,   y0+1, z0+1), c, border_value);
+    In s111 = NN.at(ivec3(x0+1, y0+1, z0+1), c, border_value);
+    float qx = x - x0;
+    float px = 1 - qx;
+    float qy = y - y0;
+    float py = 1 - qy;
+    float qz = z - z0;
+
+    float s00 = s000 * px + s001 * qx;
```

Comment:
As you pointed out during our discussion, this could be written as 
`s000 + (s001-s000) * qx` to have one multiplication instead of two

Reply:
It turned out it isn't faster - but less accurate.